### ISSUE TITLE
Fix deadlocks in SQL Server when multiple connected servers are updating message statistics for a channel

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3334]
A similar solution to what was done to resolve these deadlocks for a single server may work. It would involve inserting a dummy row with null metadata_id and null server_id. All updates would need to update that row first to prevent deadlocks from occurring.

However it may not work with tables with existing data since the dummy row would not be the first row inserted to the table. One way to resolve that problem would be to remove a re-insert all records on migration with the dummy row inserted first.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3433
Reporter: brentm
Created: 2014-09-03T16:59:23.000-0700

**Comment by rbeckman-nextgen** [comment]
Rather than adding a row with null metadata_id and null server_id, the statements that update message statistics are now overridden in SQL Server to force a table lock to be acquired. This prevents multiple servers from updating stats for a given channel at the same time and prevents the deadlocks from occurring.

Imported Comment. Original Details:
Author: brentm
Created: 2014-09-17T15:29:47.000-0700

**Comment by rbeckman-nextgen** [comment]
Replicated the issue in 3.0.3 by connecting two server and admin instances to the same database with a JS Reader as the source connector, returning a messaging with a polling frequency of 1 millisecond. Almost instantly I received a bunch of errors, one of which was a deadlock resource error.

I performed the same steps using the latest build of 3.1, connecting two server instances to the same database, dispatching messages through the same channel. This time, I let the channels run for some time. When I checked the dashboard for both admins, no deadlock errors were reported by log4j.

Imported Comment. Original Details:
Author: jaysenp
Created: 2014-09-18T12:29:03.000-0700
